service/networking.py

```diff
@@ -1,8 +1,303 @@
+
+    def get_or_create_router_interface(self, network_driver, user_neutron, router, subnet, interface_name):
+        interface = network_driver.add_router_interface(
+            router, subnet, interface_name)
+        return interface
+
+    #NOTE: Reversed order for deletes.
+    def delete_router_interface(self, network_driver, user_neutron,
+                                router_name, subnet_name):
+        try:
+            interface = network_driver.remove_router_interface(
+                network_driver.neutron, router_name, subnet_name)
+        except NotFound:
+            #This is OKAY!
+            return None
+        except:
+            raise
+        return interface
+
+    def delete_router_gateway(self, network_driver, user_neutron, router_name):
+        return network_driver.remove_router_gateway(router_name)
+
+    def delete_router(self, network_driver, user_neutron, router_name):
+        return network_driver.delete_router(user_neutron, router_name)
+
+    def delete_network(self, network_driver, user_neutron, network_name):
+        return network_driver.delete_network(user_neutron, network_name)
+
+    def delete_subnet(self, network_driver, user_neutron, subnet_name):
+        """
+        NOTE: If you see errors like the one below when you attempt to delete
+        the users network, and no instances remain, you are likely
+        not removing the router interface.
+        ```
+            Conflict: Unable to complete operation on subnet 1-2-3-4.
+                      One or more ports have an IP allocation from this subnet.
+        ```
+        """
+        return network_driver.delete_subnet(user_neutron, subnet_name)
+
+
+class ExternalNetwork(NetworkTopology):
+    """
+    This topology assumes:
+    user_subnet --> user_router --> interface --> external_network
+    """
+    name = "External Network Topology"
+    external_network_name = None
+    user_subnet_required = True
+    user_network_required = False
+    user_router_required = True
+
+    def __init__(self, identity):
+        network_name = identity.get_credential('network_name')
+        if not network_name:
+            network_name = identity.provider.get_credential('network_name')
+        if not network_name:
+            raise Exception("Unknown Network - Identity %s is missing 'network_name' " % identity)
+        self.external_network_name = network_name
+
+    def validate(self, core_identity):
+        identity_creds = core_identity.get_all_credentials()
+        if 'network_name' not in identity_creds.keys():
+            logger.warn("Credential 'network_name' missing:"
+                        "cannot create user network")
+            raise Exception("Identity %s has not been assigned a 'network_name'" % core_identity)
+        return True
+
+    def delete_network(self, network_driver, user_neutron, network_name):
+        return None
+
+    def get_or_create_network(self, network_driver, user_neutron, network_name):
```

Comment:
No. In an "External Network" topology failing to have the external network name results in an incomplete/invalid topology. Atmosphere cannot "Externalize a network".